### PR TITLE
Add slash command mentions

### PR DESCRIPTION
**Pull request by freya022**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines][contributing].

### Changes

- [X] Internal code
- [X] Library interface (affecting end-user code) 
- [X] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

Adds slash command mentions, as can be seen [here](https://discord.com/developers/docs/reference#message-formatting)

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/internal/interactions/command/CommandImpl.java [hunk 1]
```diff
@@ -221,6 +221,14 @@ public long getIdLong()
         return id;
     }
 
+    @Nonnull
+    @Override
+    public String getAsMention()
+    {
+        Checks.check(getType() == Type.SLASH, "Only slash commands can be mentioned");
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/interactions/command/CommandImpl.java:228]: This shouldn't use `Checks.check` and instead use `throw new IllegalStateException`. Checks is only for arguments.

File: src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public String toString()
+    {
+        return String.format("SlashCommandReference: name=%s, id=%d", name, id);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java:54]: Use instanceof instead

File: src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        SlashCommandReference that = (SlashCommandReference) o;
+        return id == that.id;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return (int) (id ^ (id >>> 32));
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java:62]: Should use `Long.hashCode` but also check that the name is the same. A subcommand mention uses the same id but has a different name, which should be refelected by equals and hashcode.

File: src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public String getName()
+    {
+        return name;
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("SlashCommandReference: name=%s, id=%d", name, id);
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/interactions/command/SlashCommandReference.java:47]: This isn't how we want to format string representations, see https://github.com/DV8FromTheWorld/JDA/pull/1965#issuecomment-1029608355

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * Represents a mentionable slash command.
+ */
+public interface ICommandReference extends IMentionable
+{
+    /**
+     * Returns the name of the slash command
+     *
+     * @return the name of the slash command
+     */
+    String getName();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:33]: Missing nullability annotation

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import javax.annotation.Nonnull;
+
+/**
+ * Represents a mentionable slash command.
+ */
+public interface ICommandReference extends IMentionable
+{
+    /**
+     * Returns the name of the slash command
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:29]: Might also include the subcommand name

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     *
+     * @return the name of the slash command
+     */
+    String getName();
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p><b>This will only work on slash commands !</b>
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:38]:
```suggestion
     * <p><b>This will only work on slash commands!</b>
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+     *
+     * @throws IllegalStateException
+     *         If the command is not a slash command (i.e. not of type {@link Command.Type#SLASH}
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:41]:
```suggestion
     *         If the command is not a slash command (i.e. not of type {@link Command.Type#SLASH})
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/Command.java [hunk 9]
```diff
@@ -48,7 +48,7 @@
  * @see Guild#retrieveCommandById(String)
  * @see Guild#retrieveCommands()
  */
-public interface Command extends ISnowflake
+public interface Command extends ISnowflake, ICommandReference
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/Command.java:51]: Subcommands should implement this too

**Review by MinnDevelopment — CHANGES_REQUESTED** [review]

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 10]
```diff
@@ -1990,7 +1990,13 @@ default MessageCreateAction replyFiles(@Nonnull Collection<? extends FileUpload>
...
          * Represents a mention for all users in a server, literal {@code @everyone}.
          */
-        EVERYONE("@everyone", "everyone");
+        EVERYONE("@everyone", "everyone"),
+        /**
+         * Represents a mention for a slash command.
+         * <br>The first group is the command name, the second group is the subcommand group (nullable),
+         * the third group is the subcommand (nullable), and the fourth group is the command ID.
+         */
+        SLASH_COMMAND("</([\\w-]+)(?> ([\\w-]+))??(?> ([\\w-]+))?:(\\d+)>", null);
```
Comment (Tais993) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java:1999]: Don't know what Minn and DV8 think about this, but I feel that group names are a lot more readable
```suggestion
        SLASH_COMMAND("</(?<commandName>[\\w-]+)(?<subcommandGroup>> ([\\w-]+))??(?<subcommandName>> ([\\w-]+))?:(?<commandId>\\d+)>", null);
```

Comment (freya022) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java:1999]: Other regexes there don't do that

File: src/main/java/net/dv8tion/jda/api/entities/Mentions.java [hunk 11]
```diff
@@ -325,6 +326,48 @@
...
+     * @return Immutable list of mentioned slash commands, or an empty list
+     */
+    @Nonnull
+    List<SlashCommandReference> getSlashCommands();
+
+    /**
+     * A {@link org.apache.commons.collections4.Bag Bag} of mentioned {@link SlashCommandReference slash commands}.
+     * <br>This can be used to retrieve the amount of times a slash commands was mentioned.
+     *
+     * <p>Be aware these mentions could be not mentioning an actual command
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/api/entities/Mentions.java:344]:
```suggestion
     * <p>Be aware these mentions could be mentioning a non-existent command
```

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 12]
```diff
@@ -1990,7 +1990,13 @@ default MessageCreateAction replyFiles(@Nonnull Collection<? extends FileUpload>
         /**
          * Represents a mention for all users in a server, literal {@code @everyone}.
          */
-        EVERYONE("@everyone", "everyone");
+        EVERYONE("@everyone", "everyone"),
+        /**
+         * Represents a mention for a slash command.
+         * <br>The first group is the command name, the second group is the subcommand group (nullable),
+         * the third group is the subcommand (nullable), and the fourth group is the command ID.
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java:1997]:
```suggestion
         * <br>The first group is the command name, the second group is the subcommand group name (nullable),
         * the third group is the subcommand name (nullable), and the fourth group is the command ID.
```

File: src/main/java/net/dv8tion/jda/api/entities/Message.java (hunk 10 above)
Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java:1999]: As a note, the other regexes are preeeetty basic which is why they don't do this. there's no big reason to not do it 🤷

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Override
+    public int hashCode()
+    {
+        int result = Long.hashCode(id);
+        result = 31 * result + name.hashCode();
+        result = 31 * result + (subcommand != null ? subcommand.hashCode() : 0);
+        result = 31 * result + (subcommandGroup != null ? subcommandGroup.hashCode() : 0);
+        return result;
+    }
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:119]: Probably can simplify this to use `Object.hashCode(...)`

File: src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java [hunk 14]
```diff
@@ -255,9 +277,18 @@ public boolean isMentioned(@Nonnull IMentionable mentionable, @Nonnull Message.M
                     return true;
                 break;
             case EMOJI:
-                if (mentionable instanceof Emoji && getCustomEmojis().contains(mentionable))
+                if (mentionable instanceof CustomEmoji && getCustomEmojis().contains(mentionable))
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java:280]: This seems unrelated? Do we actually want to make this change?

File: src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java [hunk 15]
```diff
@@ -255,9 +277,18 @@ public boolean isMentioned(@Nonnull IMentionable mentionable, @Nonnull Message.M
...
                     return true;
                 break;
+            case SLASH_COMMAND:
+                if (mentionable instanceof ICommandReference)
+                {
+                    final ICommandReference reference = (ICommandReference) mentionable;
+                    for (SlashCommandReference r : getSlashCommands())
+                        if (r.getCommandPath().equals(reference.getCommandPath()) && r.getIdLong() == reference.getIdLong())
+                            return true;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java:289]: Move this to a helper method like the other cases do.

File: src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java (hunk 14 above)
Comment (freya022) [reply src/main/java/net/dv8tion/jda/internal/entities/mentions/AbstractMentions.java:280]: `#getCustomEmojis` returns a `List<CustomEmoji>`, it makes no sense to check for (possibly) unicode emojis in a list of custom emojis

**Review by DV8FromTheWorld — APPROVED** [review]

**Review by DV8FromTheWorld — APPROVED** [review]
We need to add a way to _make_ a mention without having to build the entire command too probably.
But this is good enough for me for now 👍

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+ * Represents a slash command mention, such as {@code </ban soft:1021082477038678126>}
+ */
+public class SlashCommandReference implements ICommandReference
+{
+    private final long id;
+    private final String name;
+    private final String subcommand;
+    private final String subcommandGroup;
+
+    public SlashCommandReference(String name, String subcommandGroup, String subcommand, long id)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:34]: This is missing annotations

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        if (subcommand != null)
+            joiner.add(subcommand);
+
+        return joiner.toString();
+    }
+
+    @Override
+    public String toString()
+    {
+        return "SlashCommandReference: " + getCommandPath() + " (" + getId() + ")";
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:94]: This should use the new `EntityString` class

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        if (this == o) return true;
+        if (!(o instanceof SlashCommandReference)) return false;
+
+        SlashCommandReference that = (SlashCommandReference) o;
+
+        if (id != that.id) return false;
+        if (!name.equals(that.name)) return false;
+        if (!Objects.equals(subcommand, that.subcommand)) return false;
+        return Objects.equals(subcommandGroup, that.subcommandGroup);
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:108]:
```suggestion
        return id == that.id
            && name.equals(that.name)
            && Objects.equals(subcommand, that.subcommand)
            && Objects.equals(subcommandGroup, that.subcommandGroup);
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/Command.java [hunk 19]
```diff
@@ -966,6 +999,13 @@ public LocalizationMap getNameLocalizations()
             return nameLocalizations;
         }
 
+        @Nonnull
+        @Override
+        public String getCommandPath()
+        {
+            return parentCommand.getCommandPath() + " " + getName();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/Command.java:1006]: Can you change the `getCommandPath` in [CommandInteractionPayload](https://github.com/DV8FromTheWorld/JDA/blob/7b3ba8355d98ab82a479947a38138589d70f35d1/src/main/java/net/dv8tion/jda/api/interactions/commands/CommandInteractionPayload.java#L117) to be consistent with this

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        final StringJoiner joiner = new StringJoiner(" ");
+        joiner.add(name);
+        if (subcommandGroup != null)
+            joiner.add(subcommandGroup);
+        if (subcommand != null)
+            joiner.add(subcommand);
+
+        return joiner.toString();
+    }
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:89]: I agree with this implementation, however it is in contrast to the `getCommandPath` on `CommandInteractionPayload` which uses `/` as the joiner.

Can we update the `CommandInteractionPayload` implementation (and docs) to use spaces instead?
This is ready to merge after that.

**Review by DV8FromTheWorld — APPROVED** [review]

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    @Nonnull
+    String getName();
+
+    /**
+     * Returns the entire path (separated with spaces) of this command
+     *
+     * @return the entire path (separated with spaces) of this command
+     */
+    @Nonnull
+    String getFullCommandName();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:42]: The docs for this could be more descriptive. What is separated with spaces? What does path mean? The docs in CommandInteractionPayload are more detailed than this. I'd also say it depends on whether this is used on a subcommand or not.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+     * Returns the entire path (separated with spaces) of this command, which joins the current command name, as well as the parent names.
+     *
+     * <p>Examples:
+     * <ul>
+     *     <li>You are in the subcommand {@code "ban"}, contained in the {@code "mod"} command: {@code "mod ban"}</li>
+     *     <li>You are in the subcommand {@code "owner"}, contained in the {@code "config"} group and in the {@code "admin"} command: {@code "admin config owner"}</li>
+     *     <li>You are in the command {@code "ban"}: {@code "ban"}</li>
+     * </ul>
+     *
+     * @return the entire path (separated with spaces) of this command
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:46]:
```suggestion
     * Returns the full command name, including possible subcommand name and subcommand group name.
     * <br>This is the name shown on messages or when writing the command in the text input.
     *
     * <p>Examples:
     * <ul>
     *     <li>When used on a normal base command,  the full name is the name itself, as in the name {@code "ban"} is equal to the full name {@code "ban"}.</li>
     *     <li>When used on a subcommand {@code "ban"}, of the base command {@code "mod"}, the full name resolves to {@code "mod ban"}</li>
     *     <li>When the subcommand is part of a subcommand group, {@code "action"} it resolves to {@code "mod action ban"}</li>
     * </ul>
     *
     * @return the full command name
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+
+import javax.annotation.Nonnull;
+
+/**
+ * Represents a mentionable slash command.
+ */
+public interface ICommandReference extends IMentionable
+{
+    /**
+     * Returns the name of the slash command
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/ICommandReference.java:29]:
```suggestion
     * Returns the name of the slash command.
     * <br>If used on a subcommand, it returns the subcommand name.
     * And when used on a subcommand group, it returns the group name.
     * <br>You can use {@link #getFullCommandName()} to get the full name including base command and group.
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+        return name;
+    }
+
+    /**
+     * Returns the subcommand of the slash command
+     *
+     * @return the subcommand of the slash command
+     */
+    @Nullable
+    public String getSubcommand()
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java:63]:
```suggestion
    public String getSubcommandName()
```

**Review by MinnDevelopment — APPROVED** [review]
